numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -1653,10 +1653,16 @@ _array_fromobject_generic(
     /* fast exit if simple call */
     if (PyArray_CheckExact(op) || (subok && PyArray_Check(op))) {
         oparr = (PyArrayObject *)op;
-        if (type == NULL) {
+        PyArray_Descr* dtype = PyArray_AdaptDescriptorToArray(
+            oparr, dt_info.dtype, dt_info.descr);
```

Comment:
This doesn't seem ideal.  It may do the full rodeo if figuring out the string length for:
```
arr = np.array(["asdf", "fdsa"], dtype=object)
np.array(arr, dtype="U")  # won't use fast-path, but will enter it here.
```

Maybe just skip the fast path maybe unless `NPY_DTYPE(PyArray_DESCR(oparr)) == dt_info.dtype` with a comment that this is not be perfect?

N.B.: I am getting a bit confused about the behavior for `dtype=LongDType`, say it annoyingly, is a non-native byte-order `>longlong`.  Should we force a native byte order result?  We currently do, which we want for copies, but it would be fathomable to not to.

This (maybe) continues a thought I had before:  The way we have multiple integer types is awful (I think we all agree on that :)).  One solution with a bit of "less worry about the details of what happens if we remove it" factor would be to have a class hierarchy:
```
class Int64DType:
    kind = "l"  # system dependend
    num = 7

LongDType = Int64DType  # just an alias (for this system)

class LongLongDType(Int64DType):
    kind = "q"
    num = 9
```
So that `LongLong` is a subclass.  You could still cast to it and you would still (typically) get a dtype with `kind == "q"`.  These subclasses full-fill the substitution principle I think (they behave exactly the same, only have those attribute differences).

If we ever weed out the actual ufunc loops, the ufunc might return the "preferred" version, but even now do that for *many* (not all, especially simple ones like `add`) ufuncs!
The point would be though:  If you do `np.asarray(..., dtype="q")` you would get the identical type character.

Reply:
I didn't deal with the N.B. comment. Instead of completely avoiding the fast path, I found it easiest to avoid the call to `PyArray_AdaptDescriptorToArray` in cases you were worried about. It turns out that your case already drops out of the fast path due to the call to `PyArray_EquivTypes` below.